src/main/java/org/dataone/hashstore/hashstoreconverter/HashStoreConverter.java

```diff
@@ -0,0 +1,95 @@
+package org.dataone.hashstore.hashstoreconverter;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.dataone.hashstore.ObjectMetadata;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.security.NoSuchAlgorithmException;
+import java.util.Properties;
+
+/**
+ * HashStoreConverter is a utility tool to assist with ingesting existing data objects and their
+ * respective system metadata into a HashStore. Instead of duplicating data objects (that already
+ * exists), HashStoreConverter provide a simple API to create a hard link to the data object with
+ * the expected HashStore object path.
+ */
+public class HashStoreConverter {
+    private static final Log logHashStoreConverter = LogFactory.getLog(HashStoreConverter.class);
+    private FileHashStoreLinks fileHashStoreLinks;
+
+    /**
+     * Constructor to initialize HashStoreConverter. Properties to an existing or desired HashStore
+     * are required.
+     *
+     * @param hashstoreProperties Properties object with the following keys: storePath, storeDepth,
+     *                            storeWidth, storeAlgorithm, storeMetadataNamespace
+     * @throws IOException              Issue with directories or hashstore.yaml config
+     * @throws NoSuchAlgorithmException Unsupported FileHashStoreLinks algorithm
+     */
+    public HashStoreConverter(Properties hashstoreProperties)
+        throws IOException, NoSuchAlgorithmException {
+        try {
+            fileHashStoreLinks = new FileHashStoreLinks(hashstoreProperties);
+            logHashStoreConverter.info("HashStoreConverter initialized");
+
+        } catch (IOException ioe) {
+            String errMsg = "Unexpected IOException encountered: " + ioe.getMessage();
+            logHashStoreConverter.error(errMsg);
+            throw ioe;
+
+        } catch (NoSuchAlgorithmException nsae) {
+            String errMsg = "A supplied algorithm is not supported: " + nsae.getMessage();
+            logHashStoreConverter.error(errMsg);
+            throw nsae;
+
+        }
+    }
+
+    /**
+     * Take an existing path to a data object, store it into a new or existing HashStore via a
+     * hard link (to save disk space), store the supplied system metadata and return the
+     * ObjectMetadata for the data object.
+     *
+     * @param filePath      Path to existing data object
+     * @param pid           Persistent or authority-based identifier
+     * @param sysmetaStream Stream to sysmeta content to store
+     * @return ObjectMetadata for the given pid
+     * @throws IOException              An issue with calculating checksums or storing sysmeta
+     * @throws NoSuchAlgorithmException An algorithm defined is not supported
+     * @throws InterruptedException     Issue with synchronizing storing metadata
+     */
+    public ObjectMetadata convert(Path filePath, String pid, InputStream sysmetaStream)
```

Comment:
This method doesn't handle the use case - objects only have system metadata without real bytes. I mentioned that the data objects on cn don't have bytes; they only have system metadata.

Reply:
Sorry @taojing2002 - I missed this requirement. I have refactored the `convert` method to account for this - and also swapped the order of operations to store sysmeta first (because it can never be null).